src/qdmi/Driver.cpp

```diff
@@ -0,0 +1,479 @@
+    LOAD_STATIC_SYMBOL(prefix, device_job_cancel)                              \
+    LOAD_STATIC_SYMBOL(prefix, device_job_check)                               \
+    LOAD_STATIC_SYMBOL(prefix, device_job_wait)                                \
+    LOAD_STATIC_SYMBOL(prefix, device_job_get_results)                         \
+    /* device query interface */                                               \
+    LOAD_STATIC_SYMBOL(prefix, device_session_query_device_property)           \
+    LOAD_STATIC_SYMBOL(prefix, device_session_query_site_property)             \
+    LOAD_STATIC_SYMBOL(prefix, device_session_query_operation_property)        \
+    /* NOLINTEND(cppcoreguidelines-pro-type-reinterpret-cast) */               \
+    /* initialize the device */                                                \
+    device_initialize();                                                       \
+  }                                                                            \
+                                                                               \
+  prefix##DeviceLibrary::~prefix##DeviceLibrary() {                            \
+    /* Check if QDMI_device_finalize is not NULL before calling it. */         \
+    if (device_finalize != nullptr) {                                          \
+      device_finalize();                                                       \
+    }                                                                          \
+  }
+DEFINE_STATIC_LIBRARY(MQT_NA)
+
+#ifndef _WIN32
+DynamicDeviceLibrary::DynamicDeviceLibrary(const std::string& libName,
+                                           const std::string& prefix)
+    : libHandle_(dlopen(libName.c_str(), RTLD_NOW | RTLD_LOCAL)) {
+  if (libHandle_ == nullptr) {
+    throw std::runtime_error("Couldn't open the device library: " + libName);
+  }
+  if (openLibHandles().contains(libHandle_)) {
+    // dlopen uses reference counting, so we need to decrement the reference
+    // count that was increased by dlopen.
+    dlclose(libHandle_);
+    throw std::runtime_error("Device library already loaded: " + libName);
+  }
+  openLibHandles().emplace(libHandle_);
+
+// Macro for loading a symbol from the dynamic library.
+// @param symbol is the name of the symbol to load.
+#define LOAD_DYNAMIC_SYMBOL(symbol)                                            \
+  {                                                                            \
+    const std::string symbolName = std::string(prefix) + "_QDMI_" + #symbol;   \
+    (symbol) = reinterpret_cast<decltype(symbol)>(                             \
+        dlsym(libHandle_, symbolName.c_str()));                                \
+    if ((symbol) == nullptr) {                                                 \
+      throw std::runtime_error("Failed to load symbol: " + symbolName);        \
+    }                                                                          \
+  }
```

Comment:
this could use some whitespace after the macro. 
I accidentally thought that the code after the macro actually belonged to the macro.